sklearn/tree/_criterion.pyx

```diff
@@ -445,9 +579,12 @@ cdef class ClassificationCriterion(Criterion):
                 self.weighted_n_left += w
 
         else:
+            # Reverse reset the available values alone.
+            direction = self.missing_direction
             self.reverse_reset()
+            self.move_missing(direction)
```

Comment:
(To be more clear)

For each node there are available values and missing values. The float threshold is decided based on the available values and the missing_direction is decided based on the missing values.

When the Criterion traverses the available values from left to right, this heuristic (or trick) is not used. But at the end, after searching for the best threshold for each feature, we decide the node's best threshold and best feature and set them for the node and compute it's left/right child's impurity... This is where it is helpful to use this trick of updating it from left (after a `reset`) if the `pos` (threshold) is closer to left extreme or updating it from right (after a `reverse_reset`) if the `pos` is closer to right extreme.

While doing so without changing the direction of the missing values (which are stored at the beginning of the array), I here get the current missing direction. `reset` it fully (pos = start) then fix `pos` using `move_missing` (which does not exactly move anything to avoid costs of moving)...

Hope I helped clarify...